proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+## Proposed Library Change Specification
+
+- `compactAddHollow#`, `reifyInfoTablePtr#`, and `InfoTablePtrOf#` are exposed in `GHC.Prim`
+- `LiftedCtorToSymbol` is exposed in `GHC.TypeLits`
+
+## Examples
+
+### Primitive usage
+
+#### Allocate a hollow `Left` constructor in region `compactRegion`
+
+```haskell
+let leftInfoPtr :: InfoTablePtrOf# 'Left = reifyInfoTablePtr# (# #)
+ in compactAddHollow# compactRegion (unsafeCoerceAddr leftInfoPtr) state
+```
+
+Yet unspecified fields can later be assigned with existing primop `writeAddrOffAddr#` (we know the offsets of the fields thanks to the return type of `compactAddHollow#`).
+
+#### Allocate hollow `stg_IND` object in region `compactRegion`
+
+```haskell
+let indInfoPtr :: InfoTablePtrOf# "IND" = reifyInfoTablePtr# (# #)
+ in compactAddHollow# compactRegion (unsafeCoerceAddr indInfoPtr) state
+```
+
+#### Get info about fields of a data constructor using `GHC.Generics` and `LiftedCtorToSymbol`
+
+```haskell
+type family (a :: [k]) ++ (b :: [k]) :: [k] where
+  '[] ++ y = y
+  (x : xs) ++ y = x : (xs ++ y)
+  _ ++ _ = TypeError ('Text "No match for ++")
+
+type family FromJust (x :: Maybe k) :: k where
+  FromJust ('Just v) = v
+  FromJust 'Nothing = TypeError ('Text "FromJust error: got 'Nothing")
+
+type family (x :: Maybe k) <|> (y :: Maybe k) :: Maybe k where
+  ('Just v) <|> _ = 'Just v
+  'Nothing <|> y = y
+
+type family GFieldsOf (repA :: Type) :: [(Meta, Type)] where
+  GFieldsOf (S1 meta f p) = '[ '(meta, GSlotTypeOf (f p))]
+  GFieldsOf (U1 _) = '[]
+  GFieldsOf ((f :*: g) p) = GFieldsOf (f p) ++ GFieldsOf (g p)
+  GFieldsOf (M1 _ _ f p) = GFieldsOf (f p)
+  GFieldsOf _ = TypeError ('Text "No match for GFieldsOf")
+
+type family GSlotTypeOf (repA :: Type) :: Type where
+  GSlotTypeOf (K1 _ c _) = c
+  GSlotTypeOf (M1 _ _ f p) = GSlotTypeOf (f p)
+  GSlotTypeOf _ = TypeError ('Text "No match for GSlotTypeOf")
+
+type family GSpecCtorOf (symCtor :: Symbol) (repA :: Type) :: Maybe (Meta, [(Meta, Type)]) where
+  GSpecCtorOf symCtor (C1 ('MetaCons symCtor x y) f p) = 'Just '(('MetaCons symCtor x y), GFieldsOf (f p))
+  GSpecCtorOf symCtor (C1 ('MetaCons _ _ _) _ _) = 'Nothing
+  GSpecCtorOf symCtor ((f :+: g) p) = GSpecCtorOf symCtor (f p) <|> GSpecCtorOf symCtor (g p)
+  GSpecCtorOf symCtor (V1 _) = 'Nothing
+  GSpecCtorOf symCtor (M1 _ _ f p) = GSpecCtorOf symCtor (f p)
+  GSpecCtorOf _ _ = TypeError ('Text "No match for GSpecCtorOf")
+
+type family LiftedCtorToSpecCtor lCtor (a :: Type) :: (Meta, [(Meta, Type)]) where
+  LiftedCtorToSpecCtor lCtor a = FromJust (GSpecCtorOf (LiftedCtorToSymbol lCtor) (Rep a ()))
+```
+
+This type-level machinery first use `LiftedCtorToSymbol` to get the name of the specified data constructor as a `Symbol`, and then uses it to filter the generic representation of the associated type. That way, `LiftedCtorToSpecCtor` can go from a lifted data constructor to a type-level list of its fields with their metadata and type.
+
+### Library side, simple example
```

Comment:
This is a nice example of the DPS library. But I'm not sure that it's too relevant here. None of your changes happen. The S-expression example is good because you point to the paper and explain the results, no distraction.

Reply:
I cut down the example a little. But I would like to keep it in the proposal.